### DISCUSSION TITLE
Load a new ride type from a .parkobj

### DISCUSSION THREAD

**Discussion by frutiemax** [OpenRCT2/OpenRCT2#16921] [Feature requests]
It would be very nice to be able to define a new ride type from a .parkobj file, defined in objects.json. Here are the features I would want:

- Define a custom size base for a flat ride (5x5, 6x3, 12x4, etc.). Since there are many possibilities, I'd suggest to create a TrackElementDescriptor when loading the object and save the ID in the ride type descriptor. This also implies to be able to load/save a TrackElementDescriptor from JSON.
- Save/Load the RideTypeDescriptor to JSON.
- Handle the ride operation
- Handle the ride painting
- Be able to load additional trackElements, this implies saving the IDs of the additional pieces in the RTD.

**Comment by frutiemax** [comment]
I've made a sample file for a TrackElementDescriptor embedded in a json file.

```
{
    "id": "openrct2.track_element.FlatTrack2x2",
    "authors": ["OpenRCT2 developers"],
    "version": "1.0",
    "sourceGame": "rct2",
    "objectType": "track_element",
    "properties": {
        "Description": 10,
		"Coordinates": {
			"rotation_begin" : 0,
			"rotation_end" : 0,
			"z_begin" : 0,
			"z_end" : 0,
			"x" : 0,
			"y" : 0
		}
		"Block": [
			{"index"=0, "x"=0, "y"=0, "z"=0, "var_07"=0, "var_08"={"tileQuarter"=0, "zQuarter"=0}, "flags"=0},
			{"index"=0, "x"=0, "y"=0, "z"=0, "var_07"=0, "var_08"={"tileQuarter"=0, "zQuarter"=0}, "flags"=0}
		],
		"PieceLength"= 0,
		"CurveChain"={"next"="openrct2.track_element.FlatTrack2x2", "previous"="openrct2.track_element.FlatTrack2x2"},
		"AlternateType"="openrct2.track_element.FlatTrack2x2",
		"Price"=0,
		"MirrorElement"="openrct2.track_element.FlatTrack2x2",
		"HeightMarkerPositions"=0,
		"Flags"=0,
		"SequenceElementAllowedWallEdges"=[
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
		],
		"SequenceProperties"=[
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
		],
		"Definition"={"type"=0, "vangle_end"=0, "vangle_start"=0, "bank_end"=0, "bank_start"=0, "preview_z_offset"=0},
		"SpinFunction"=0,
		"VerticalFactor"=0,
		"LateralFactor"=0
    },
    "strings": {
        "name": {
            "en-GB": "FlatTrack2x2"
        }
    }
}
```
Some problems:

- Some fields will require that the trackType identifier should be a string and not an integer... I believe it will cause some problems with the legacy types...

- var_07 and var_08 probably needs renaming also. 
- The VerticalFactor and LateralFactor are function of the current progress of the vehicle on the track element, the json should find a way to describe a function...
- I'm also not sure how to handle a TrackElement as an object in the scenario editor... What happens if there is a required element missing for a ride?

**Reply by IntelOrca** [reply]
I am not sure we should embed the track types into the ride type objects. They are static and common between most tracked rides. New ones can be added easily to the OpenRCT2 codebase. I think the ride type should be reserved to just the properties of a ride type and graphics of the track. Although given some ride types share the same track graphics, maybe that should be another object type too.

**Reply by frutiemax** [reply]
I was suggesting making TrackElements into its own .pakobj type. RideType .parkobj would reference the trackelements by their unique string id.

**Comment by frutiemax** [comment]
Some example for a RideType object i.e. the circus one:

```
{
	"id": "openrct2.ride_type.circus",
	"authors": ["OpenRCT2 developers"],
	"version": "1.0",
	"sourceGame": "rct2",
	"object_type": "ride_type",
	"properties": {
		"RideType": "circus",
		"AlternateType" : "",
		"Category" : "gentle",
		"EnabledTrackPieces": {},
		"StartTrackPiece": "flat_track_3x3",
		"TrackPaintFunction": "circus",
		"Flags": [
			"has_single_piece_station",
			"cannot_have_gaps",
			"has_load_options",
			"has_no_track",
			"vehicle_is_integral",
			"track_no_walls",
			"flat_ride",
			"in_ride",
			"has_vehicle_colours",
			"music_on_default",
			"has_entrance_exit",
			"has_single_session",
			"list_vehicles_separately"
		],
		"RideModes": [
			"circus"
		],
		"OperatingSettings": {
			"MinValue":0,
			"MaxValue":0,
			"MaxBrakesSpeed":0,
			"PoweredLiftAcceleration":0,
			"BoosterAcceleration":0,
			"BoosterSpeedFactor":0
		},
		"Naming": {
			"Name": "ride_name_circus",
			"Description" : "ride_description_circus"
		}
		"NameConvention": {
			"vehicle":"car",
			"structure":"structure",
			"station":"station"
		}
		"AvailableBreakdowns": [
			"safety_cut_out"
		],
		"Heights": {
			"MaxHeight": 0,
			"ClearanceHeight": 0,
			"VehicleZOffset": 0,
			"PlatformHeight": 0
		},
		"MaxMass": 255,
		"LiftData":{
			"SoundId":"",
			"MinimumSpeed": 5,
			"MaximumSpeed": 5
		},
		"RatingsCalculationFunction": "circus",
		"RatingsMultipliers": {
			"Excitement": 20,
			"Intensity": 10,
			"Nausea": 0
		},
		"UpkeepCosts": {
			"BaseCost": 50,
			"TrackLengthMultiplier": 1,
			"CostPerTrackPiece": 0,
			"CostPerTrain": 0,
			"CostPerCar": 0,
			"CostPerStation": 0
		},
		"BuildCosts": {
			"TrackPrice": 62.5,
			"SupportPrice": 1.0,
			"PriceEstimateMultiplier": 1
		},
		"DefaultPrices": [
			15, 0
		]
		"DefaultMusic": "",
		"PhotoItem": "photo",
		"BonusValue": 39,
		"ColourPresets": [
			"red", "blue", "yellow
		],
		"ColourPreview": [
			0, 0
		]
		"ColourKey": "ride"
	}
}
```

For now, the embedded functions should be kept hard-coded, there's a ton of work to do just for the "easy" fields.

Some logical steps to be done:

1. Add a unit test that export software the circus ride type and save it to a .parkobj and check its output (the generated .json and the expected .json)
2. Add a unit test that loads the circus .parkobj and check if the fields matches the one in C++
3. Repeat this process for other ride types so it covers all the corner cases

At the end, there should be a final test unit that loads all the rideType .parkobjs and check if the imported ride types matches the legacy ones.
